### PR TITLE
Finish implementations of vectorized 1D & 2D plans

### PR DESCRIPTION
**Pull request by ioannisPApapadopoulos**

### CODE REVIEW

**Comment by ioannisPApapadopoulos** [comment]
@dlfivefifty I finished the implementation for the vectorized 1D and 2D plans. The tests are passing. Are you happy for me to merge and bump the package version number?

**Comment by dlfivefifty** [comment]
Can you get the code cov checks passing?

**Comment by ioannisPApapadopoulos** [comment]
ok done @dlfivefifty

**Comment by ioannisPApapadopoulos** [comment]
oh! actually sorry! there's a bug when there's more than two slices in 2D. I'll fix it

File: src/arrowhead.jl [hunk 1]
```diff
@@ -206,7 +206,7 @@ function materialize!(M::MatMulVecAdd{<:ArrowheadLayouts,<:AbstractStridedLayout
     for k = 1:d
         mul!(view(y, m+k:d:size(y,1)), A.D[k], view(x, n+k:d:size(x,1)), α, one(α))
     end
-    y_in
+    return y_in
```
Comment (dlfivefifty) [inline src/arrowhead.jl:209]: Why are you adding `return` everywhere? It's unnecessary.

**Comment by dlfivefifty** [comment]
Remove WIP from the title if its ready

File: src/arrowhead.jl (hunk 1 above)
Comment (ioannisPApapadopoulos) [reply src/arrowhead.jl:209]: The CodeCov checks fail without the return. That's what the CodeCov was complaining about earlier, my tests had covered all the new code.

**Comment by ioannisPApapadopoulos** [comment]
@dlfivefifty The performance in 2D is not as good as I was expecting. For instance:

```
using PiecewiseOrthogonalPolynomials
using ClassicalOrthogonalPolynomials
import PiecewiseOrthogonalPolynomials: BlockedArray, BlockedOneTo
P = ContinuousPolynomial{0}(range(0,1,5))
F = plan_transform(P, Block(20,20));
G = plan_transform(P, (Block(20), Block(20), 20), (1,2));

n, p = 4, 20
X = BlockedArray{Float64}(undef, (BlockedOneTo(n:n:n*p), BlockedOneTo(n:n:n*p), 1:20))
X[:,:,:] = rand(size(X)...)

@time G*(G \ X); # 0.399731 seconds (397 allocations: 234.827 MiB, 43.38% gc time)
@time for k in axes(X,3)
    F*(F \ X[:,:,k])
end # 0.202041 seconds (3.08 k allocations: 232.389 MiB, 5.28% gc time)
```
Then the tensored transform is actually slower than the for loop. I am not expert but the gc time of the tensor transform seems quite high. Any suggestions?

**Comment by dlfivefifty** [comment]
That's weird. Can you do a profile to see whats up?

**Comment by dlfivefifty** [comment]
Note the current Toeplitz–Hankel transforms in FastTransforms.jl do a very dumb thing and allocate a large tensor for memory. This can be fixed, but actually the LibFastTransforms implementation of Legendre should be getting a fast plan so we can just use that. @MikaelSlevinsky any updates?

**Comment by ioannisPApapadopoulos** [comment]
I'm not sure what I can glean from the profile.. 

`@profile (G \ X) `

gives

```
julia> Profile.print()
Overhead ╎ [+additional indent] Count File:Line; Function
=========================================================
   ╎192 @VSCodeServer\src\eval.jl:34; (::VSCodeServer.var"#64#65")()
   ╎ 192 @Base\essentials.jl:889; invokelatest(::Any)
   ╎  192 @Base\essentials.jl:892; #invokelatest#2
   ╎   192 @VSCodeServer\src\eval.jl:263; (::VSCodeServer.var"#66#71"{VSCodeServer.ReplRunCodeRequestParams})()
   ╎    192 @Base\logging.jl:627; with_logger
   ╎     192 @Base\logging.jl:515; with_logstate(f::Function, logstate::Any)
   ╎    ╎ 192 @VSCodeServer\src\eval.jl:150; (::VSCodeServer.var"#67#72"{Bool, Bool, Bool, Module, String, Int64, Int64, String, VSCodeServer.ReplRunCodeRequestParams})()
   ╎    ╎  192 @VSCodeServer\src\repl.jl:38; hideprompt(f::VSCodeServer.var"#68#73"{Bool, Bool, Bool, Module, String, Int64, Int64, String, VSCodeServer.ReplRunCodeRequestParams})
   ╎    ╎   192 @VSCodeServer\src\eval.jl:179; (::VSCodeServer.var"#68#73"{Bool, Bool, Bool, Module, String, Int64, Int64, String, VSCodeServer.ReplRunCodeRequestParams})()
   ╎    ╎    192 @VSCodeServer\src\repl.jl:276; withpath(f::VSCodeServer.var"#69#74"{Bool, Bool, Bool, Module, String, Int64, Int64, String, VSCodeServer.ReplRunCodeRequestParams}, path::String)
   ╎    ╎     192 @VSCodeServer\src\eval.jl:181; (::VSCodeServer.var"#69#74"{Bool, Bool, Bool, Module, String, Int64, Int64, String, VSCodeServer.ReplRunCodeRequestParams})()
   ╎    ╎    ╎ 192 @VSCodeServer\src\eval.jl:268; kwcall(::@NamedTuple{softscope::Bool}, ::typeof(VSCodeServer.inlineeval), m::Module, code::String, code_line::Int64, code_column::Int64, file::String)
   ╎    ╎    ╎  192 @VSCodeServer\src\eval.jl:271; inlineeval(m::Module, code::String, code_line::Int64, code_column::Int64, file::String; softscope::Bool)
   ╎    ╎    ╎   192 @Base\essentials.jl:889; invokelatest(::Any, ::Any, ::Vararg{Any})
   ╎    ╎    ╎    192 @Base\essentials.jl:892; invokelatest(::Any, ::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, @NamedTuple{}})
   ╎    ╎    ╎     192 @Base\loading.jl:2076; include_string(mapexpr::typeof(REPL.softscope), mod::Module, code::String, filename::String)
   ╎    ╎    ╎    ╎ 192 @Base\boot.jl:385; eval
   ╎    ╎    ╎    ╎  192 @PiecewiseOrthogonalPolynomials\src\piecewisepolynomial.jl:142; \(F::PiecewiseOrthogonalPolynomials.ApplyPlan{Float64, typeof(PiecewiseOrthogonalPolynomials._perm_blockvec), ClassicalOrthogonalPolynomials.JacobiT…
   ╎    ╎    ╎    ╎   4   @PiecewiseOrthogonalPolynomials\src\piecewisepolynomial.jl:137; _inv_perm_blockvec(X::BlockedArray{Float64, 3, Array{Float64, 3}, Tuple{BlockedOneTo{Int64, StepRange{Int64, Int64}}, BlockedOneTo{Int64, StepRange…
   ╎    ╎    ╎    ╎    4   @PiecewiseOrthogonalPolynomials\src\piecewisepolynomial.jl:102; _inv_perm_blockvec
   ╎    ╎    ╎    ╎     3   @PiecewiseOrthogonalPolynomials\src\piecewisepolynomial.jl:106; _inv_perm_blockvec(X::BlockArrays.BlockedMatrix{Float64, Matrix{Float64}, Tuple{BlockedOneTo{Int64, StepRange{Int64, Int64}}, BlockedOneTo{Int64,…
   ╎    ╎    ╎    ╎    ╎ 3   @Base\baseext.jl:30; Array
   ╎    ╎    ╎    ╎    ╎  3   @Base\boot.jl:494; Array
  2╎    ╎    ╎    ╎    ╎   3   @Base\boot.jl:489; Array
   ╎    ╎    ╎    ╎     1   @PiecewiseOrthogonalPolynomials\src\piecewisepolynomial.jl:108; _inv_perm_blockvec(X::BlockArrays.BlockedMatrix{Float64, Matrix{Float64}, Tuple{BlockedOneTo{Int64, StepRange{Int64, Int64}}, BlockedOneTo{Int64,…
   ╎    ╎    ╎    ╎    ╎ 1   @BlockArrays\src\blockaxis.jl:7; getindex
   ╎    ╎    ╎    ╎    ╎  1   @BlockArrays\src\blockedarray.jl:254; getindex
   ╎    ╎    ╎    ╎    ╎   1   @BlockArrays\src\blockedarray.jl:250; _blockedindex_getindex
   ╎    ╎    ╎    ╎    ╎    1   @Base\broadcast.jl:903; materialize
   ╎    ╎    ╎    ╎    ╎     1   @Base\broadcast.jl:1118; copy
   ╎    ╎    ╎    ╎    ╎    ╎ 1   @Base\ntuple.jl:49; ntuple
   ╎    ╎    ╎    ╎    ╎    ╎  1   @Base\broadcast.jl:1118; #31
   ╎    ╎    ╎    ╎    ╎    ╎   1   @Base\broadcast.jl:682; _broadcast_getindex
   ╎    ╎    ╎    ╎    ╎    ╎    1   @Base\broadcast.jl:709; _broadcast_getindex_evalf
   ╎    ╎    ╎    ╎    ╎    ╎     1   @BlockArrays\src\blockaxis.jl:4; getindex
   ╎    ╎    ╎    ╎    ╎    ╎    ╎ 1   @BlockArrays\src\blockaxis.jl:366; getindex
   ╎    ╎    ╎    ╎    ╎    ╎    ╎  1   @BlockArrays\src\blockaxis.jl:326; blockaxes
   ╎    ╎    ╎    ╎    ╎    ╎    ╎   1   @BlockArrays\src\blockaxis.jl:297; blockaxes
   ╎    ╎    ╎    ╎    ╎    ╎    ╎    1   @BlockArrays\src\blockaxis.jl:298; _blockaxes
   ╎    ╎    ╎    ╎    ╎    ╎    ╎     1   @Base\abstractarray.jl:77; axes
   ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎ 1   @Base\range.jl:706; axes
  1╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎  1   @Base\range.jl:0; length(r::StepRange{Int64, Int64})
   ╎    ╎    ╎    ╎   188 @ClassicalOrthogonalPolynomials\src\classical\jacobi.jl:83; \
   ╎    ╎    ╎    ╎    187 @AbstractFFTs\src\definitions.jl:245; \(p::FastTransforms.ChebyshevToLegendrePlanTH{Float64, FastTransforms.ToeplitzHankelPlan{Float64, 5, 6, Tuple{Matrix{Float64}, Matrix{Float64}}, Tuple{FastTransforms.Toepli…
   ╎    ╎    ╎    ╎     37  @FastTransforms\src\toeplitzhankel.jl:52; *
   ╎    ╎    ╎    ╎    ╎ 3   @FastTransforms\src\toeplitzhankel.jl:43; _th_applymul!(d::Int64, v::Array{Float64, 5}, T::FastTransforms.ToeplitzPlan{Float64, 6, Int64, ComplexF64, Tuple{Vector{ComplexF64}}, FFTW.cFFTWPlan{ComplexF64, -1, …
   ╎    ╎    ╎    ╎    ╎  3   @Base\broadcast.jl:911; materialize!
   ╎    ╎    ╎    ╎    ╎   3   @Base\broadcast.jl:914; materialize!
   ╎    ╎    ╎    ╎    ╎    3   @Base\broadcast.jl:956; copyto!
   ╎    ╎    ╎    ╎    ╎     3   @Base\broadcast.jl:1003; copyto!
  1╎    ╎    ╎    ╎    ╎    ╎ 1   @Base\simdloop.jl:75; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎ 2   @Base\simdloop.jl:77; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎  2   @Base\broadcast.jl:1004; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎   2   @Base\broadcast.jl:636; getindex
   ╎    ╎    ╎    ╎    ╎    ╎    2   @Base\broadcast.jl:681; _broadcast_getindex
   ╎    ╎    ╎    ╎    ╎    ╎     2   @Base\broadcast.jl:705; _getindex
   ╎    ╎    ╎    ╎    ╎    ╎    ╎ 2   @Base\broadcast.jl:675; _broadcast_getindex
   ╎    ╎    ╎    ╎    ╎    ╎    ╎  2   @Base\multidimensional.jl:696; getindex
  2╎    ╎    ╎    ╎    ╎    ╎    ╎   2   @Base\essentials.jl:14; getindex
   ╎    ╎    ╎    ╎    ╎ 31  @FastTransforms\src\toeplitzhankel.jl:44; _th_applymul!(d::Int64, v::Array{Float64, 5}, T::FastTransforms.ToeplitzPlan{Float64, 6, Int64, ComplexF64, Tuple{Vector{ComplexF64}}, FFTW.cFFTWPlan{ComplexF64, -1, …
   ╎    ╎    ╎    ╎    ╎  4   @FastTransforms\src\toeplitzplans.jl:59; *(A::FastTransforms.ToeplitzPlan{Float64, 6, Int64, ComplexF64, Tuple{Vector{ComplexF64}}, FFTW.cFFTWPlan{ComplexF64, -1, true, 6, Int64}, AbstractFFTs.ScaledPlan{Com…
   ╎    ╎    ╎    ╎    ╎   4   @Base\array.jl:395; fill!
  4╎    ╎    ╎    ╎    ╎    4   @Base\array.jl:1021; setindex!
   ╎    ╎    ╎    ╎    ╎  5   @FastTransforms\src\toeplitzplans.jl:60; *(A::FastTransforms.ToeplitzPlan{Float64, 6, Int64, ComplexF64, Tuple{Vector{ComplexF64}}, FFTW.cFFTWPlan{ComplexF64, -1, true, 6, Int64}, AbstractFFTs.ScaledPlan{Com…
   ╎    ╎    ╎    ╎    ╎   5   @Base\abstractarray.jl:1068; copyto!(dest::SubArray{ComplexF64, 6, Array{ComplexF64, 6}, NTuple{6, Base.OneTo{Int64}}, false}, src::SubArray{Float64, 6, Array{Float64, 6}, NTuple{6, Base.OneTo{Int64}}, fals…
   ╎    ╎    ╎    ╎    ╎    5   @Base\abstractarray.jl:1102; copyto_unaliased!(deststyle::IndexCartesian, dest::SubArray{ComplexF64, 6, Array{ComplexF64, 6}, NTuple{6, Base.OneTo{Int64}}, false}, srcstyle::IndexCartesian, src::SubArray{F…
   ╎    ╎    ╎    ╎    ╎     3   @Base\abstractarray.jl:1291; getindex
   ╎    ╎    ╎    ╎    ╎    ╎ 3   @Base\abstractarray.jl:1341; _getindex
   ╎    ╎    ╎    ╎    ╎    ╎  3   @Base\subarray.jl:290; getindex
  3╎    ╎    ╎    ╎    ╎    ╎   3   @Base\essentials.jl:14; getindex
   ╎    ╎    ╎    ╎    ╎     2   @Base\abstractarray.jl:1396; setindex!
   ╎    ╎    ╎    ╎    ╎    ╎ 2   @Base\abstractarray.jl:1426; _setindex!
   ╎    ╎    ╎    ╎    ╎    ╎  2   @Base\subarray.jl:331; setindex!
  2╎    ╎    ╎    ╎    ╎    ╎   2   @Base\array.jl:1024; setindex!
   ╎    ╎    ╎    ╎    ╎  5   @FastTransforms\src\toeplitzplans.jl:63; *(A::FastTransforms.ToeplitzPlan{Float64, 6, Int64, ComplexF64, Tuple{Vector{ComplexF64}}, FFTW.cFFTWPlan{ComplexF64, -1, true, 6, Int64}, AbstractFFTs.ScaledPlan{Com…
   ╎    ╎    ╎    ╎    ╎   5   @FFTW\src\fft.jl:835; *
  5╎    ╎    ╎    ╎    ╎    5   @FFTW\src\fft.jl:518; unsafe_execute!
   ╎    ╎    ╎    ╎    ╎  5   @FastTransforms\src\toeplitzplans.jl:68; *(A::FastTransforms.ToeplitzPlan{Float64, 6, Int64, ComplexF64, Tuple{Vector{ComplexF64}}, FFTW.cFFTWPlan{ComplexF64, -1, true, 6, Int64}, AbstractFFTs.ScaledPlan{Com…
   ╎    ╎    ╎    ╎    ╎   5   @FastTransforms\src\chebyshevtransform.jl:66; applydim!(op!::Function, X::Array{ComplexF64, 6}, d::Int64, ind::Function)
   ╎    ╎    ╎    ╎    ╎    5   @FastTransforms\src\chebyshevtransform.jl:59; applydim!(op!::FastTransforms.var"#245#246"{Vector{ComplexF64}}, X::Array{ComplexF64, 6}, Rpre::CartesianIndices{0, Tuple{}}, Rpost::CartesianIndices{5, NTuple…
   ╎    ╎    ╎    ╎    ╎     5   @FastTransforms\src\toeplitzplans.jl:68; (::FastTransforms.var"#245#246"{Vector{ComplexF64}})(v::SubArray{ComplexF64, 1, Array{ComplexF64, 6}, Tuple{Base.Slice{Base.OneTo{Int64}}, Vararg{Int64, 5}}, true})
   ╎    ╎    ╎    ╎    ╎    ╎ 5   @Base\broadcast.jl:911; materialize!
   ╎    ╎    ╎    ╎    ╎    ╎  5   @Base\broadcast.jl:914; materialize!
   ╎    ╎    ╎    ╎    ╎    ╎   5   @Base\broadcast.jl:956; copyto!
   ╎    ╎    ╎    ╎    ╎    ╎    5   @Base\broadcast.jl:1003; copyto!
   ╎    ╎    ╎    ╎    ╎    ╎     5   @Base\simdloop.jl:77; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎    ╎ 5   @Base\broadcast.jl:1004; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎    ╎  5   @Base\broadcast.jl:636; getindex
   ╎    ╎    ╎    ╎    ╎    ╎    ╎   5   @Base\broadcast.jl:682; _broadcast_getindex
   ╎    ╎    ╎    ╎    ╎    ╎    ╎    5   @Base\broadcast.jl:709; _broadcast_getindex_evalf
   ╎    ╎    ╎    ╎    ╎    ╎    ╎     5   @Base\complex.jl:296; *
  5╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎ 5   @Base\float.jl:411; *
   ╎    ╎    ╎    ╎    ╎  9   @FastTransforms\src\toeplitzplans.jl:72; *(A::FastTransforms.ToeplitzPlan{Float64, 6, Int64, ComplexF64, Tuple{Vector{ComplexF64}}, FFTW.cFFTWPlan{ComplexF64, -1, true, 6, Int64}, AbstractFFTs.ScaledPlan{Com…
   ╎    ╎    ╎    ╎    ╎   9   @AbstractFFTs\src\definitions.jl:269; *(p::AbstractFFTs.ScaledPlan{ComplexF64, FFTW.cFFTWPlan{ComplexF64, 1, true, 6, Int64}, Float64}, x::Array{ComplexF64, 6})
   ╎    ╎    ╎    ╎    ╎    4   @FFTW\src\fft.jl:835; *
  4╎    ╎    ╎    ╎    ╎     4   @FFTW\src\fft.jl:518; unsafe_execute!
   ╎    ╎    ╎    ╎    ╎    5   C:\workdir\usr\share\julia\stdlib\v1.10\LinearAlgebra\src\generic.jl:182; rmul!
   ╎    ╎    ╎    ╎    ╎     4   @Base\simdloop.jl:77; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎ 4   C:\workdir\usr\share\julia\stdlib\v1.10\LinearAlgebra\src\generic.jl:183; macro expansion
  4╎    ╎    ╎    ╎    ╎    ╎  4   @Base\essentials.jl:13; getindex
   ╎    ╎    ╎    ╎    ╎     1   @Base\simdloop.jl:78; macro expansion
  1╎    ╎    ╎    ╎    ╎    ╎ 1   @Base\int.jl:87; +
   ╎    ╎    ╎    ╎    ╎  3   @FastTransforms\src\toeplitzplans.jl:74; *(A::FastTransforms.ToeplitzPlan{Float64, 6, Int64, ComplexF64, Tuple{Vector{ComplexF64}}, FFTW.cFFTWPlan{ComplexF64, -1, true, 6, Int64}, AbstractFFTs.ScaledPlan{Com…
   ╎    ╎    ╎    ╎    ╎   3   @Base\broadcast.jl:911; materialize!
   ╎    ╎    ╎    ╎    ╎    3   @Base\broadcast.jl:914; materialize!
   ╎    ╎    ╎    ╎    ╎     3   @Base\broadcast.jl:956; copyto!
   ╎    ╎    ╎    ╎    ╎    ╎ 3   @Base\broadcast.jl:1003; copyto!
   ╎    ╎    ╎    ╎    ╎    ╎  3   @Base\simdloop.jl:77; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎   3   @Base\broadcast.jl:1004; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎    3   @Base\abstractarray.jl:1396; setindex!
   ╎    ╎    ╎    ╎    ╎    ╎     3   @Base\abstractarray.jl:1426; _setindex!
   ╎    ╎    ╎    ╎    ╎    ╎    ╎ 3   @Base\subarray.jl:331; setindex!
  3╎    ╎    ╎    ╎    ╎    ╎    ╎  3   @Base\array.jl:1024; setindex!
   ╎    ╎    ╎    ╎    ╎ 2   @FastTransforms\src\toeplitzhankel.jl:45; _th_applymul!(d::Int64, v::Array{Float64, 5}, T::FastTransforms.ToeplitzPlan{Float64, 6, Int64, ComplexF64, Tuple{Vector{ComplexF64}}, FFTW.cFFTWPlan{ComplexF64, -1, …
   ╎    ╎    ╎    ╎    ╎  2   @Base\broadcast.jl:911; materialize!
   ╎    ╎    ╎    ╎    ╎   2   @Base\broadcast.jl:914; materialize!
   ╎    ╎    ╎    ╎    ╎    2   @Base\broadcast.jl:956; copyto!
   ╎    ╎    ╎    ╎    ╎     2   @Base\broadcast.jl:1003; copyto!
  1╎    ╎    ╎    ╎    ╎    ╎ 1   @Base\simdloop.jl:75; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎ 1   @Base\simdloop.jl:77; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎  1   @Base\broadcast.jl:1004; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎   1   @Base\broadcast.jl:636; getindex
   ╎    ╎    ╎    ╎    ╎    ╎    1   @Base\broadcast.jl:682; _broadcast_getindex
   ╎    ╎    ╎    ╎    ╎    ╎     1   @Base\broadcast.jl:709; _broadcast_getindex_evalf
  1╎    ╎    ╎    ╎    ╎    ╎    ╎ 1   @Base\float.jl:411; *
   ╎    ╎    ╎    ╎    ╎ 1   @FastTransforms\src\toeplitzhankel.jl:46; _th_applymul!(d::Int64, v::Array{Float64, 5}, T::FastTransforms.ToeplitzPlan{Float64, 6, Int64, ComplexF64, Tuple{Vector{ComplexF64}}, FFTW.cFFTWPlan{ComplexF64, -1, …
   ╎    ╎    ╎    ╎    ╎  1   @Base\reducedim.jl:1036; sum!
   ╎    ╎    ╎    ╎    ╎   1   @Base\reducedim.jl:1036; #sum!#853
   ╎    ╎    ╎    ╎    ╎    1   @Base\reducedim.jl:1034; sum!
   ╎    ╎    ╎    ╎    ╎     1   @Base\reducedim.jl:1034; #sum!#852
   ╎    ╎    ╎    ╎    ╎    ╎ 1   @Base\reducedim.jl:324; mapreducedim!
   ╎    ╎    ╎    ╎    ╎    ╎  1   @Base\reducedim.jl:316; _mapreducedim!(f::typeof(identity), op::typeof(Base.add_sum), R::Array{Float64, 5}, A::SubArray{Float64, 6, Array{Float64, 6}, NTuple{6, Base.OneTo{Int64}}, false})
   ╎    ╎    ╎    ╎    ╎    ╎   1   @Base\simdloop.jl:77; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎    1   @Base\reducedim.jl:317; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎     1   @Base\multidimensional.jl:698; setindex!
  1╎    ╎    ╎    ╎    ╎    ╎    ╎ 1   @Base\array.jl:1024; setindex!
   ╎    ╎    ╎    ╎     150 @FastTransforms\src\toeplitzhankel.jl:170; inv
   ╎    ╎    ╎    ╎    ╎ 150 @FastTransforms\src\toeplitzhankel.jl:203; plan_th_leg2cheb!
   ╎    ╎    ╎    ╎    ╎  150 @Base\broadcast.jl:903; materialize
   ╎    ╎    ╎    ╎    ╎   150 @Base\broadcast.jl:1118; copy
   ╎    ╎    ╎    ╎    ╎    150 @Base\ntuple.jl:49; ntuple
   ╎    ╎    ╎    ╎    ╎     150 @Base\broadcast.jl:1118; #31
   ╎    ╎    ╎    ╎    ╎    ╎ 150 @Base\broadcast.jl:692; _broadcast_getindex
   ╎    ╎    ╎    ╎    ╎    ╎  150 @Base\broadcast.jl:709; _broadcast_getindex_evalf
   ╎    ╎    ╎    ╎    ╎    ╎   150 @FastTransforms\src\toeplitzhankel.jl:179; _leg2chebTH_TLC(::Type{Float64}, mn::NTuple{5, Int64}, d::Int64)
   ╎    ╎    ╎    ╎    ╎    ╎    37  @FastTransforms\src\toeplitzplans.jl:104; plan_uppertoeplitz!(v::Vector{Float64}, szs::NTuple{6, Int64}, dim::Int64)
   ╎    ╎    ╎    ╎    ╎    ╎     37  @Base\array.jl:632; zeros
   ╎    ╎    ╎    ╎    ╎    ╎    ╎ 28  @Base\array.jl:636; zeros
 28╎    ╎    ╎    ╎    ╎    ╎    ╎  28  @Base\boot.jl:489; Array
   ╎    ╎    ╎    ╎    ╎    ╎    ╎ 9   @Base\array.jl:637; zeros
   ╎    ╎    ╎    ╎    ╎    ╎    ╎  6   @Base\array.jl:395; fill!
  6╎    ╎    ╎    ╎    ╎    ╎    ╎   6   @Base\array.jl:1021; setindex!
  3╎    ╎    ╎    ╎    ╎    ╎    ╎  3   @Base\array.jl:396; fill!
   ╎    ╎    ╎    ╎    ╎    ╎    113 @FastTransforms\src\toeplitzplans.jl:106; plan_uppertoeplitz!(v::Vector{Float64}, szs::NTuple{6, Int64}, dim::Int64)
   ╎    ╎    ╎    ╎    ╎    ╎     113 @Base\abstractarray.jl:829; similar
   ╎    ╎    ╎    ╎    ╎    ╎    ╎ 113 @Base\abstractarray.jl:830; similar
   ╎    ╎    ╎    ╎    ╎    ╎    ╎  113 @Base\array.jl:420; similar
113╎    ╎    ╎    ╎    ╎    ╎    ╎   113 @Base\boot.jl:489; Array
   ╎    ╎    ╎    ╎    1   @FastTransforms\src\chebyshevtransform.jl:204; \(P::FastTransforms.ChebyshevTransformPlan{Float64, 1, Vector{Int32}, false, 5, Tuple{Int64, Int64}}, x::Array{Float64, 5})
   ╎    ╎    ╎    ╎     1   @FastTransforms\src\chebyshevtransform.jl:325; *
   ╎    ╎    ╎    ╎    ╎ 1   @FastTransforms\src\chebyshevtransform.jl:267; mul!(y::Array{Float64, 5}, P::FastTransforms.IChebyshevTransformPlan{Float64, 1, Vector{Int32}, false, 5, Tuple{Int64, Int64}}, x::Array{Float64, 5})
   ╎    ╎    ╎    ╎    ╎  1   @FastTransforms\src\chebyshevtransform.jl:238; _icheb1_prescale!
   ╎    ╎    ╎    ╎    ╎   1   @FastTransforms\src\chebyshevtransform.jl:233; _icheb1_prescale!
   ╎    ╎    ╎    ╎    ╎    1   @FastTransforms\src\chebyshevtransform.jl:77; lmul_dim_begin!
   ╎    ╎    ╎    ╎    ╎     1   @FastTransforms\src\chebyshevtransform.jl:66; applydim!(op!::Function, X::Array{Float64, 5}, d::Int64, ind::Int64)
   ╎    ╎    ╎    ╎    ╎    ╎ 1   @FastTransforms\src\chebyshevtransform.jl:59; applydim!(op!::FastTransforms.var"#178#180"{Int64}, X::Array{Float64, 5}, Rpre::CartesianIndices{0, Tuple{}}, Rpost::CartesianIndices{4, NTuple{4, Base.OneTo…
   ╎    ╎    ╎    ╎    ╎    ╎  1   @FastTransforms\src\chebyshevtransform.jl:77; #178
   ╎    ╎    ╎    ╎    ╎    ╎   1   C:\workdir\usr\share\julia\stdlib\v1.10\LinearAlgebra\src\generic.jl:220; lmul!
   ╎    ╎    ╎    ╎    ╎    ╎    1   @Base\simdloop.jl:77; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎     1   C:\workdir\usr\share\julia\stdlib\v1.10\LinearAlgebra\src\generic.jl:221; macro expansion
   ╎    ╎    ╎    ╎    ╎    ╎    ╎ 1   @Base\subarray.jl:343; setindex!
  1╎    ╎    ╎    ╎    ╎    ╎    ╎  1   @Base\array.jl:1021; setindex!
Total snapshots: 192. Utilization: 100% across all threads and tasks. Use the `groupby` kwarg to break down by thread and/or task.
```

**Comment by TSGut** [comment]
You can use something like https://github.com/timholy/ProfileView.jl to visualize where your code is spending most of its time. I personally prefer visualized profiling but I guess that block of text also contains the same info.

**Comment by ioannisPApapadopoulos** [comment]
prof

**Comment by dlfivefifty** [comment]
Why would `\` itself be calling `plan_th_leg2cheb!`?

**Comment by ioannisPApapadopoulos** [comment]
That actually happens independently of PiecewiseOrthogonalPolynomials.jl. E.g. that call also happens if one runs

```
F = plan_transform(Legendre(), 10_000, 1)
c = rand(10_000)
F \ c;
```

**Comment by dlfivefifty** [comment]
Probably you want `plan_itransform` then

**Comment by ioannisPApapadopoulos** [comment]
Hm sure. But `*` is also slower. E.g.

```
using PiecewiseOrthogonalPolynomials
using ClassicalOrthogonalPolynomials
import PiecewiseOrthogonalPolynomials: BlockedArray, BlockedOneTo
n, p, nz = 4, 20, 20
P = ContinuousPolynomial{0}(range(0,1,n+1))
F = plan_transform(P, Block(p,p));
G = plan_transform(P, (Block(p), Block(p), nz), (1,2));

X = BlockedArray{Float64}(undef, (BlockedOneTo(n:n:n*p), BlockedOneTo(n:n:n*p), 1:nz))
X[:,:,:] = rand(size(X)...)
Xp = G \ X;
@time G*Xp; # 0.116234 seconds (174 allocations: 3.939 MiB)
@time for k in axes(Xp,5)
    F*Xp[:,:,:,:,k]
end # 0.103507 seconds (1.78 k allocations: 3.006 MiB)
```

Flame graph:

image

File: src/arrowhead.jl (hunk 1 above)
Comment (dlfivefifty) [reply src/arrowhead.jl:209]: Oh weird. I didn’t know adding return fixed the bug with codecov

annoying but whatevs

**Pull request merged**